### PR TITLE
Parallel helper for all addons to use.

### PR DESCRIPTION
**Pull request by BlueWeabo**
Features:

- Calculates Parallels for you!
- Cleaner looking code :D
- Batch Mode (Optional)
- Void Protection (Optional and currently experimental) - @skizzerz code
- Automatic Consumption of inputs
- Automatic Output Calculation

### CODE REVIEW

File: src/main/java/gregtech/api/util/GT_ParallelHelper.java [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+            }
+
+            for (GT_MetaTileEntity_Hatch tHatch : mMachineMeta.mOutputHatches) {
+                if (tHatch instanceof GT_MetaTileEntity_Hatch_Output_ME) {
+                    tMEOutputHatch = true;
+                    break;
+                }
+            }
+
+            if (!tMEOutputBus || !tMEOutputHatch) {
```
Comment (minecraft7771) [inline src/main/java/gregtech/api/util/GT_ParallelHelper.java:259]: Remove this outer if, its unnecessary.

**Review by minecraft7771 — CHANGES_REQUESTED** [review]
Other than minor style complaint this looks good to me and I'm looking forward to using it in machines.

File: src/main/java/gregtech/api/util/GT_ParallelHelper.java (hunk 1 above)
Comment (BlueWeabo) [reply src/main/java/gregtech/api/util/GT_ParallelHelper.java:259]: left over code is a thing i forget to remove. this one if went through multiple iterations kek

**Review by minecraft7771 — APPROVED** [review]

File: src/main/java/gregtech/api/util/GT_ParallelHelper.java [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+            // now that all partial stacks have been counted, create a priority queue for our outputs
+            // the lowest priority item is the number of complete parallel crafts we can support
+            PriorityQueue<MutableTriple<Integer, Integer, ItemStack>> aParallelQueue =
+                    new PriorityQueue<>(Comparator.comparing(MutableTriple::getLeft));
+            for (Entry<ItemStack, MutablePair<Integer, Integer>> entry : tParallels.entrySet()) {
+                aParallelQueue.add(new MutableTriple<>(entry.getValue().left, entry.getValue().right, entry.getKey()));
+            }
+
+            while (tSlotsFree > 0) {
+                MutableTriple<Integer, Integer, ItemStack> tParallel = aParallelQueue.poll();
```
Comment (Glease) [inline src/main/java/gregtech/api/util/GT_ParallelHelper.java:465]: I have slight doubt over this.. what if output hatches aren't the same size? or should we just disregard this possibility whatsoever?

Comment (BlueWeabo) [reply src/main/java/gregtech/api/util/GT_ParallelHelper.java:465]: Did you ask the question for items? Aka we caring that an output bus will have different sizes?

Comment (Glease) [reply src/main/java/gregtech/api/util/GT_ParallelHelper.java:465]: 😅commented on the wrong section of code. I meant the fluid output hatches. Not item. sorry.

Comment (BlueWeabo) [reply src/main/java/gregtech/api/util/GT_ParallelHelper.java:465]: I am not really understanding the question...

Comment (BlueWeabo) [reply src/main/java/gregtech/api/util/GT_ParallelHelper.java:465]: Talking through the phone on GitHub is a bit hard haha. But we are accounting for the size of the hatch for fluids. Also realized I have fucked up some comments lol

**Pull request merged**